gusto/form_manipulation_labelling.py

```diff
@@ -0,0 +1,328 @@
+import ufl
+import functools
+import operator
+from firedrake import Function, TrialFunction, Constant
+from firedrake.formmanipulation import split_form
+
+
+identity = lambda t: t
+drop = lambda t: None
+all_terms = lambda t: True
+extract = lambda idx: lambda t: Term(split_form(t.form)[idx].form, t.labels)
+
+
+def has_labels(*labels):
+    """:arg labels: the :class:`Label` object(s)
+
+    Returns a function that takes in a :class:`Term` object and checks
+    whether is has the label(s).
+    """
+
+    def t_has_labels(t):
+        if len(labels) == 1:
+            return t.has_label(labels[0])
+        else:
+            return any(t.has_label(*labels))
+
+    return t_has_labels
+
+
+def linearise(t):
+    """
+    :arg t: a :class:`.Term` object
+
+    Returns a :class:`.LabelledForm` object.
+
+    The linearisation of t must have been specified by calling the
+    :class:`.linearisation` on t. Returns a labelled form with terms that
+    have the linear form from l=t.get("linearisation") and inherit labels
+    from both t and l, with the labels from l overwriting those from t if
+    both present.
+    """
+    t_linear = t.get(linearisation)
+    assert type(t_linear) == LabelledForm
+    new_labels = t.labels.copy()
+    new_labels["linearisation"] = True
+    return LabelledForm(functools.reduce(
+        operator.add,
+        [Term(l.form, dict(new_labels, **l.labels))
+         for l in t_linear]))
+
+
+def replace_test(new_test):
+    """
+    :arg new_test: a :func:`TestFunction`
+
+    Returns a function that takes in t, a :class:`Term`, and returns
+    a new :class:`Term` with form containing the new_test and
+    labels=t.labels
+    """
+
+    def rep(t):
+        test = t.form.arguments()[0]
+        new_form = ufl.replace(t.form, {test: new_test})
+        return Term(new_form, t.labels)
```

Comment:
Does this work if `new_test` and `test` are `TestFunctions` from a mixed finite element space? What about more general replacements like replacing test functions on a mixed space of two spaces `W = U * V` with ones in `Wtilde = U * V * T`?

Reply:
I haven't tried... in fact I make sure that we assert that the function space isn't mixed before we use this (see line 240 of advection.py). At the moment this function is only used in the `CrankNicolson` timestepper where we advect the fields separately. At the moment I am only trying to replicate the functionality that we already have. I imagine when I want to apply an IMEX scheme to the mixed problem, I will have to revisit this....